README.md
# css-commons

[![Build Status](https://travis-ci.org/gabrielcsapo/css-commons.svg?branch=master)](https://travis-ci.org/gabrielcsapo/css-commons)
[![Dependency Status](https://starbuck.gabrielcsapo.com/badge/github/gabrielcsapo/css-commons/status.svg)](https://starbuck.gabrielcsapo.com/github/gabrielcsapo/css-commons)
[![devDependency Status](https://starbuck.gabrielcsapo.com/badge/github/gabrielcsapo/css-commons/dev-status.svg)](https://starbuck.gabrielcsapo.com/github/gabrielcsapo/css-commons#info=devDependencies)
[![Coverage Status](https://lcov-server.gabrielcsapo.com/badge/github%2Ecom/gabrielcsapo/css-commons.svg)](https://lcov-server.gabrielcsapo.com/coverage/github%2Ecom/gabrielcsapo/css-commons)
[![npm](https://img.shields.io/npm/dt/css-commons.svg?maxAge=2592000)]()
[![npm](https://img.shields.io/npm/dm/css-commons.svg?maxAge=2592000)]()


> 👨‍👨‍👧‍👧 let's make our css a family again

`css-commons` aims to reduce the number of orphaned selectors that only contain one rule that is similar to other selectors rules.

<!-- TOC depthFrom:1 depthTo:6 withLinks:1 updateOnSave:1 orderedList:0 -->

- [css-commons](#css-commons)
	- [Installation](#installation)
	- [Usage](#usage)

<!-- /TOC -->

## Installation

```
npm install css-commons -g
```

## Usage

> `css-commons` is a commonjs module and can be placed as a script inside a page using `./dist/css-commons.js` (132k) or `./dist/css-commons.min.js` (42k)

```
cat main.css | css-commons > main.min.css
```

```js
const compile = require('css-commons')

const css = `
  .alert.alert-black * {
    color: #fff;
  }
  .alert.alert-default {
    background-color: #cfcfc4;
    border: 1px solid #bdbdae;
  }
  .alert.alert-default * {
    color: #fff;
  }
  .badge.badge-white {
    background-color: #fff;
    color: #000;
  }
  .badge.border-white {
    color: #000;
  }
  .badge.badge-black {
    background-color: #000;
    color: #fff;
  }
  .badge.border-black {
    color: #000;
  }
  pre {
    display: block;
    padding: 10px;
    font-size: 13px;
    line-height: 1.42857143;
    word-break: break-all;
    word-wrap: break-word;
    white-space: normal;
    background-color: #f5f5f5;
    border-radius: 5px;
    border-left: 0.3rem solid transparent;
  }
  .grid > .col-1-12 {
    width: 8.333333333333332%;
    margin-left: 0%;
    margin-right: 0%;
  }
  @media screen and (min-width: 32em) {
    .grid > .col-sm-0-12 {
      display: none;
    }
    .grid > .col-sm-1-12 {
      width: 8.3333%;
      margin-left: 0%;
      margin-right: 0%;
    }
  }
`;
var output = commons(css);
```

> the above snippet will render the resulting css

```css
@media screen and (min-width: 32em) {
  .grid>.col-sm-0-12 {
    display: none;
  }
  .grid>.col-sm-1-12 {
    width: 8.3333%;
    margin-left: 0%;
    margin-right: 0%;
  }
}

.alert.alert-black *, .alert.alert-default *, .badge.badge-black {
  color: #fff;
}

.alert.alert-default {
  background-color: #cfcfc4;
  border: 1px solid #bdbdae;
}

.badge.badge-white {
  background-color: #fff;
}

.badge.badge-white, .badge.border-white, .badge.border-black {
  color: #000;
}

.badge.badge-black {
  background-color: #000;
}

pre {
  display: block;
  padding: 10px;
  font-size: 13px;
  line-height: 1.42857143;
  word-break: break-all;
  word-wrap: break-word;
  white-space: normal;
  background-color: #f5f5f5;
  border-radius: 5px;
  border-left: 0.3rem solid transparent;
}

.grid>.col-1-12 {
  width: 8.333333333333332%;
  margin-left: 0%;
  margin-right: 0%;
}
```

> The rules that contained `color: #fff;` or `color: #000;` were combined.

> This resulted in a bundle size that weighs in an at `847B` which used to be `1206B`, this is a 29% decrease in overall size.


CHANGELOG.md
# 0.0.4 (11/24/2017)

- removes the need for the dedent module
- updates dependencies

# 0.0.3 (10/22/2017)

- does two passes on the data to ensure optimizations are kept and created

# 0.0.2 (10/22/2017)

- adds bin location in package.json

# 0.0.1 (10/22/2017)

- basic use cases work
- includes commonjs module


test/index.js
const test = require('tape');

const commons = require('../');

test('css-commons', (t) => {
	t.plan(2);

	t.test('should be able to combine children', (t) => {
		var css = `
      .alert.alert-black * {
        color: #fff;
      }
      .alert.alert-default {
        background-color: #cfcfc4;
        border: 1px solid #bdbdae;
      }
      .alert.alert-default * {
        color: #fff;
      }
      .badge.badge-white {
        background-color: #fff;
        color: #000;
      }
      .badge.border-white {
        color: #000;
      }
      .badge.badge-black {
        background-color: #000;
        color: #fff;
      }
      .badge.border-black {
        color: #000;
      }
      pre {
        display: block;
        padding: 10px;
        font-size: 13px;
        line-height: 1.42857143;
        word-break: break-all;
        word-wrap: break-word;
        white-space: normal;
        background-color: #f5f5f5;
        border-radius: 5px;
        border-left: 0.3rem solid transparent;
      }
      .grid > .col-1-12 {
        width: 8.333333333333332%;
        margin-left: 0%;
        margin-right: 0%;
      }
      @media screen and (min-width: 32em) {
        .grid > .col-sm-0-12 {
          display: none;
        }
        .grid > .col-sm-1-12 {
          width: 8.3333%;
          margin-left: 0%;
          margin-right: 0%;
        }
      }
    `;
		var output = commons(css);
		t.deepEqual(output, '@media screen and (min-width: 32em) {\n  .grid > .col-sm-0-12 {\n    display: none;\n  }\n\n  .grid > .col-sm-1-12 {\n    width: 8.3333%;\n    margin-left: 0%;\n    margin-right: 0%;\n  }\n}\n.alert.alert-black *,.alert.alert-default *,.badge.badge-black {\n  color:#fff; \n}\n.alert.alert-default {\n  background-color:#cfcfc4;\n  border:1px solid #bdbdae; \n}\n.badge.badge-white {\n  background-color:#fff; \n}\n.badge.badge-white,.badge.border-white,.badge.border-black {\n  color:#000; \n}\n.badge.badge-black {\n  background-color:#000; \n}\npre {\n  display:block;\n  padding:10px;\n  font-size:13px;\n  line-height:1.42857143;\n  word-break:break-all;\n  word-wrap:break-word;\n  white-space:normal;\n  background-color:#f5f5f5;\n  border-radius:5px;\n  border-left:0.3rem solid transparent; \n}\n.grid > .col-1-12 {\n  width:8.333333333333332%;\n  margin-left:0%;\n  margin-right:0%; \n}\n');
		t.equal(output.length, 854);
		t.equal(css.length, 1206);
		t.end();
	});

	t.test('shouldn\'t break up already existing optimizations', (t) => {
		const css = `
			.tooltip:before,
			[data-tooltip]:before {
			  z-index: 1001;
			  border: 6px solid transparent;
			  background: transparent;
			  content: '';
			}
		`;
		t.equal(commons(css), '.tooltip:before,[data-tooltip]:before {\n  z-index:1001;\n  border:6px solid transparent;\n  background:transparent;\n  content:\'\'; \n}\n');
		t.end();
	});

});


index.js
/** @module css-commons */

const css = require('css');

/**
 * combines orphaned and similar selectors with others
 * @param  {string} original - the original css string
 * @return {string} - the mutated css string
 * @example
const commons = require('css-commons');
const css = `
 .alert.alert-black * {
	 color: #fff;
 }
 .alert.alert-default {
	 background-color: #cfcfc4;
	 border: 1px solid #bdbdae;
 }
 .alert.alert-default * {
	 color: #fff;
 }
 .badge.badge-white {
	 background-color: #fff;
	 color: #000;
 }
 .badge.border-white {
	 color: #000;
 }
 .badge.badge-black {
	 background-color: #000;
	 color: #fff;
 }
 .badge.border-black {
	 color: #000;
 }
 pre {
	 display: block;
	 padding: 10px;
	 font-size: 13px;
	 line-height: 1.42857143;
	 word-break: break-all;
	 word-wrap: break-word;
	 white-space: normal;
	 background-color: #f5f5f5;
	 border-radius: 5px;
	 border-left: 0.3rem solid transparent;
 }
`;
var output = commons(css);

 */
module.exports = (original) => {
	const obj = css.parse(original, {});

	let declartions = {};
	let pass = {};
	let output = '';

	obj.stylesheet.rules.forEach((rule) => {
		if(!rule.declarations) {
			output += css.stringify({
				'type': 'stylesheet',
				'stylesheet': {
					'rules': [rule]
				}
			}) + '\n';
			return;
		}
		rule.declarations.forEach((decl) => {
			var key = `${decl.property}:${decl.value}`;
			if(declartions[key]) {
				declartions[key] = declartions[key].concat(rule.selectors);
			} else {
				declartions[key] = rule.selectors;
			}
		});
	});

	Object.keys(declartions).forEach((s) => {
		// if there is only one child, stack these for later
		if(declartions[s].length === 1) {
			if(pass[declartions[s][0]]) {
				pass[declartions[s][0]].push(s);
			} else {
				pass[declartions[s][0]] = [s];
			}
		} else {
			// optimize by doing a second pass and collating selectors
			const key = `${declartions[s].join(',')}`;
			if(pass[key]) {
				pass[key] = pass[key].concat([s]);
			} else {
				pass[key] = [s];
			}
		}
	});

	// now output the final value
	Object.keys(pass).forEach((s) => {
		output += `${s} {\n\u00A0 ${pass[s].join(';\n\u00A0\u00A0')}; \n}` + '\n';
	});

	return output;
};


tryitout.js
const pkg = require('./package.json');

module.exports = {
	title: pkg.name,
	description: pkg.description,
	nav: {
		Source: 'https://github.com/gabrielcsapo/css-commons'
	},
	body: [{
		type: 'text',
		value: '> [`css-commons`](https://github.com/gabrielcsapo/css-commons) aims to reduce the number of orphaned selectors that only contain one rule that is similar to other selectors rules.'
	}, {
		type: 'code',
		title: 'removed orphan selectors',
		value: `
    const commons = require('css-commons');

    const css = \`
      .alert.alert-black * {
        color: #fff;
      }
      .alert.alert-default {
        background-color: #cfcfc4;
        border: 1px solid #bdbdae;
      }
        .alert.alert-default * {
        color: #fff;
      }
      .badge.badge-white {
        background-color: #fff;
        color: #000;
      }
      .badge.border-white {
        color: #000;
      }
      .badge.badge-black {
        background-color: #000;
        color: #fff;
      }
      .badge.border-black {
        color: #000;
      }
      pre {
        display: block;
        padding: 10px;
        font-size: 13px;
        line-height: 1.42857143;
        word-break: break-all;
        word-wrap: break-word;
        white-space: normal;
        background-color: #f5f5f5;
        border-radius: 5px;
        border-left: 0.3rem solid transparent;
      }
      .grid > .col-1-12 {
        width: 8.333333333333332%;
        margin-left: 0%;
        margin-right: 0%;
      }
      @media screen and (min-width: 32em) {
        .grid > .col-sm-0-12 {
          display: none;
        }
        .grid > .col-sm-1-12 {
          width: 8.3333%;
          margin-left: 0%;
          margin-right: 0%;
        }
      }
    \`;

    console.log(commons(css));
  `
	}, {
		type: 'text',
		value: `
  > As you can see after pressing run, the rules that contained \`color: #fff;\` or \`color: #000;\` were combined.

  > This resulted in a bundle size that weighs in an at \`847B\` which used to be \`1206B\` this is a \`29%\` decrease in overall size.
  `
	}],
	output: './docs',
	externals: [
		'./dist/css-commons.js'
	]
};
